modules/post/windows/manage/vss.rb

```diff
@@ -0,0 +1,128 @@
+        'Platform' => ['win'],
+        'SessionTypes' => ['meterpreter'],
+        'Author' => ['theLightCosine'],
+        'References' => [
+          [ 'URL', 'http://pauldotcom.com/2011/11/safely-dumping-hashes-from-liv.html' ]
+        ],
+        'Actions' => [
+          [ 'VSS_CREATE', { 'Description' => 'Create a new VSS copy' } ],
+          [ 'VSS_LIST_COPIES', { 'Description' => 'List VSS copies' } ],
+          [ 'VSS_MOUNT', { 'Description' => 'Mount a VSS copy' } ],
+          [ 'VSS_UNMOUNT', { 'Description' => 'Unmount a VSS copy' } ],
+          [ 'VSS_GET_INFO', { 'Description' => 'Get VSS information' } ],
+          [ 'VSS_SET_MAX_STORAGE_SIZE', { 'Description' => 'Set the VSS maximum storage size' } ]
+        ],
+        'DefaultAction' => 'VSS_GET_INFO'
+      )
+    )
+
+    register_options(
+      [
+        OptInt.new('SIZE', [ false, 'Size in bytes to set for max storage.' ], conditions: %w[ACTION == VSS_SET_MAX_STORAGE_SIZE]),
+        OptString.new('VOLUME', [ false, 'Volume to make a copy of.', 'C:\\' ], conditions: %w[ACTION == VSS_CREATE]),
+        OptString.new('DEVICE', [ false, 'DeviceObject of the shadow copy to mount.' ], conditions: %w[ACTION == VSS_MOUNT]),
+        OptString.new('PATH', [ false, 'Path to use for mounting the shadow copy.', 'ShadowCopy' ], conditions: ['ACTION', 'in', %w[VSS_MOUNT VSS_UNMOUNT] ])
+      ]
+    )
+  end
+
+  def run
+    # all conditional options are required when active, make sure none of them are blank
+    options.each_pair do |name, option|
+      next if option.conditions.empty?
+      next unless Msf::OptCondition.show_option(self, option)
+
+      fail_with(Failure::BadConfig, "The #{name} option is required by the #{action.name} action.") if datastore[name].blank?
+    end
+
+    fail_with(Failure::NoAccess, 'This module requires administrative privileges to run') unless is_admin?
+    fail_with(Failure::NoAccess, 'This module requires UAC to be bypassed first') if is_uac_enabled?
+    fail_with(Failure::Unknown, 'Failed to start the necessary VSS services') unless start_vss
+
+    send("action_#{action.name.downcase}")
+  end
+
+  def action_vss_create
+    if (id = create_shadowcopy(datastore['VOLUME']))
+      print_good "Shadow Copy #{id} created!"
+    end
+  end
+
+  def action_vss_get_info
+    return unless (storage_data = vss_get_storage)
+
+    tbl = Rex::Text::Table.new(
+      'Header' => 'Shadow Copy Storage Data',
+      'Indent' => 2,
+      'Columns' => ['Field', 'Value']
+    )
+    storage_data.each_pair { |k, v| tbl << [k, v] }
+    print_good(tbl.to_s)
+    store_loot('host.shadowstorage', 'text/plain', session, tbl.to_s, 'shadowstorage.txt', 'Shadow Copy Storage Info')
+  end
+
+  def action_vss_mount
+    print_status('Creating the symlink...')
+    device = datastore['DEVICE']
+    device << '/' unless device.end_with?('/') # the DEVICE parameter needs to end with / or the link will be created successfully but will not work
```

Comment:
I noticed that adding `/` only works if the DEVICE path uses slashes too. If it uses backslashes, it has to end with `\`. Apparently, they cannot be mixed up:
```
//?/GLOBALROOT/Device/HarddiskVolumeShadowCopy1/
```
or
```
\\?\GLOBALROOT\Device\HarddiskVolumeShadowCopy1\
```
but not
```
\\?\GLOBALROOT\Device\HarddiskVolumeShadowCopy1/
```

Reply:
That's a really interesting observation. In commit 104a9575d82b3f805707a60e1a147e2328c6d1ad I switched to using a regex to perform validation on the `DEVICE` parameter. I intentionally left it pretty loose, mostly validating the prefix since I'm not sure all of the legitimate cases. It does however validate that only one type of slash is used and then appends that type of slash to the end as necessary. I thought checking it at runtime and failing with `BadConfig` would be better than using the `regex` datastore keyword argument since the error message will be more clear. This will also hopefully help users realize which value they should be using from the `vss_list_copies` output.